rfcs/text/0012_background_sessions.md

```diff
@@ -0,0 +1,291 @@
+// this should return quickly.
+const response$ = await searchService.search({
+  params,
+  sessionId,
+  options,
+});
+
+// When the user decides to send to background and move on, create the saved object associated with this session.
+// Make sure to store any app state needed to restore this view and the exact search requests from before.
+searchService.session.save({
+  appState,
+  restoreUrl,
+});
+```
+
+# Motivation
+
+Kibana is great at providing fast results from large sets of "hot" data. However, there is an increasing number of use
+cases where cost is more of a concern than speed (such as year-over-year reports, historical or audit data, batch
+queries, etc.). When running slower searches on "cold" storage or frozen indices, users encounter timeouts after 30
+seconds by default.
+
+In 7.7, with the introduction of the `_async_search` API in Elasticsearch, we provided Kibana users a way to bypass the
+timeout, but users still need to remain on-screen for the entire duration of the search requests.
+
+By default, when a user navigates away before search requests finish, we cancel them. This is a good default behavior,
+since there currently isn't any way for the user to later return to the results.
+
+The primary motivation of this RFC is to enable users to do the following without needing to keep Kibana open, or while
+moving onto other work inside Kibana:
+
+- Run long search requests (beyond 30 seconds)
+- View their status (complete/incomplete)
+- Cancel incomplete search requests
+- Retrieve completed search request results 
+
+# Detailed design
+
+Because a single view (such as a dashboard with multiple visualizations) can initiate multiple search requests, we need
+a way to associate the search requests together in a single entity. We call this entity a `session`, and when a user
+decides that they want to continue running the search requests while moving onto other work, we will create a saved
+object corresponding with that specific `session`.
+
+We will expose a new `session` service inside of the `data` plugin `search` service. The service will expose APIs for 
+the following:
+
+- Starting a new session (generating a unique identifier for this specific grouping of search requests)
+- Saving a session (creating the saved object using that ID)
+- Getting a list of sessions for the current user (including status, any necessary application data, and a URL to
+restore the view)
+- Cancelling an incomplete session (deleting the saved object and cancelling any incomplete search requests)
+- Restoring a session (allowing search requests to resume instead of restart)
+
+We will also modify the existing `search` API exposed by the `data` plugin `search` service to optionally accept an
+additional parameter, the `sessionId`, which will allow search requests to *resume* instead of *restart*. When
+"restoring" a session, it is crucial that search requests are sent with *identical* search request parameters as were
```

Comment:
After a session has been restored. If the user was to do something to alter one of the queries, for example by changing a filter just for one visualization (this might not be possible now), how will this work for end-users?